gnovm/pkg/doc/doc.go

```diff
@@ -0,0 +1,303 @@
+	if o == nil {
+		o = &WriteDocumentationOptions{}
+	}
+	o.w = w
+
+	var err error
+	// pkgData may already be initialised if we already had to look to see
+	// if it had the symbol we wanted; otherwise initialise it now.
+	if d.pkgData == nil {
+		d.pkgData, err = newPkgData(d.Dir, o.Unexported)
+		if err != nil {
+			return err
+		}
+	}
+
+	astpkg, pkg, err := d.pkgData.docPackage(o)
+	if err != nil {
+		return err
+	}
+
+	// copied from go source - map vars, constants and constructors to their respective types.
+	typedValue := make(map[*doc.Value]bool)
+	constructor := make(map[*doc.Func]bool)
+	for _, typ := range pkg.Types {
+		pkg.Consts = append(pkg.Consts, typ.Consts...)
+		pkg.Vars = append(pkg.Vars, typ.Vars...)
+		pkg.Funcs = append(pkg.Funcs, typ.Funcs...)
+		if o.Unexported || token.IsExported(typ.Name) {
+			for _, value := range typ.Consts {
+				typedValue[value] = true
+			}
+			for _, value := range typ.Vars {
+				typedValue[value] = true
+			}
+			for _, fun := range typ.Funcs {
+				// We don't count it as a constructor bound to the type
+				// if the type itself is not exported.
+				constructor[fun] = true
+			}
+		}
+	}
+
+	pp := &pkgPrinter{
+		name:        d.pkgData.name,
+		pkg:         astpkg,
+		file:        ast.MergePackageFiles(astpkg, 0),
+		doc:         pkg,
+		typedValue:  typedValue,
+		constructor: constructor,
+		fs:          d.pkgData.fset,
+		opt:         o,
+		importPath:  d.importPath,
+	}
+	pp.buf.pkg = pp
+
+	return d.output(pp)
+}
+
+func (d *documentable) output(pp *pkgPrinter) (err error) {
+	defer func() {
+		pp.flush()
+		if err == nil {
+			err = pp.err
+		}
+	}()
+
+	switch {
+	case d.symbol == "" && d.accessible == "":
+		if pp.opt.ShowAll {
+			pp.allDoc()
+			return
+		}
+		pp.packageDoc()
+	case d.symbol != "" && d.accessible == "":
+		pp.symbolDoc(d.symbol)
+	default: // both non-empty
+		if pp.methodDoc(d.symbol, d.accessible) {
+			return
+		}
+		if pp.fieldDoc(d.symbol, d.accessible) {
+			return
+		}
+	}
+
+	return
+}
+
+// set as a variable so it can be changed by testing.
+var fpAbs = filepath.Abs
+
+// ResolveDocumentable returns a Documentable from the given arguments.
+// Refer to the documentation of gnodev doc for the formats accepted (in general
+// the same as the go doc command).
+func ResolveDocumentable(dirs *Dirs, args []string, unexported bool) (Documentable, error) {
+	parsed, ok := parseArgs(args)
+	if !ok {
+		return nil, fmt.Errorf("commands/doc: invalid arguments: %v", args)
+	}
+	return resolveDocumentable(dirs, parsed, unexported)
+}
+
+func resolveDocumentable(dirs *Dirs, parsed docArgs, unexported bool) (Documentable, error) {
+	var candidates []Dir
+
+	// if we have a candidate package name, search dirs for a dir that matches it.
+	// prefer directories whose import path match precisely the package
+	if s, err := os.Stat(parsed.pkg); err == nil && s.IsDir() {
+		// expand to full path
+		absVal, err := fpAbs(parsed.pkg)
```

Comment:
swallowed error?

Reply:
I think if we have an error here it should be a warning, and to me the cases where filepath.Abs returns an error are already in the realm of "you specifically got yourself in this situation", because errors trickle down from `os.Getcwd` not working on unix and `syscall.FullPath` not working on windows. Do you think it should still be logged as a warning?

Reply:
If we don't want to warn to the user because the warning might be misleading, we should comment the code explaining that behavior and move the happy path to be the main code path:

```go
		absVal, err := fpAbs(parsed.pkg)
		if err != nil {
                        // comment explaining why this behavior
			break
		}
		
		candidates = dirs.findDir(absVal)
```